contracts/evm/contracts/adapters/ClusterConnection.sol

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+pragma abicoder v2;
+
+import "openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+import "@xcall/utils/Types.sol";
+import "@xcall/contracts/xcall/interfaces/IConnection.sol";
+import "@iconfoundation/xcall-solidity-library/interfaces/ICallService.sol";
+
+contract ClusterConnection is Initializable, IConnection {
+    mapping(string => uint256) private messageFees;
+    mapping(string => uint256) private responseFees;
+    mapping(string => mapping(uint256 => bool)) receipts;
+    mapping(address => bool) public isValidator;
+
+    address private xCall;
+    address private adminAddress;
+    uint256 public connSn;
+    address[] private validators;
+    uint8 private reqValidatorCnt;
+
+    event Message(string targetNetwork, uint256 sn, bytes _msg);
+    event ValidatorAdded(address _validator);
+    event ValidatorRemoved(address _validator);
+
+    modifier onlyAdmin() {
+        require(msg.sender == this.admin(), "OnlyRelayer");
+        _;
+    }
+
+    function initialize(address _relayer, address _xCall) public initializer {
+        xCall = _xCall;
+        validators.push(_relayer);
+        adminAddress = _relayer;
+        emit ValidatorAdded(_relayer);
+    }
+
+    function listValidators() external view returns (address[] memory) {
+        return validators;
+    }
+
+    function addValidator(address _validator) external onlyAdmin {
+        require(!isValidator[_validator], "Address is already an signer");
+        validators.push(_validator);
+        isValidator[_validator] = true;
+        emit ValidatorAdded(_validator);
+    }
+
+    function removeValidator(address _validator) external onlyAdmin {
+        require(_validator!=this.admin(), "Cannot remove admin");
+        require(isValidator[_validator],"Validator doesn't exist");
+        for (uint i = 0; i < validators.length; i++) {
+            if (validators[i] == _validator) {
+                validators[i] = validators[validators.length - 1]; 
+                validators.pop();                 
+                isValidator[_validator] = false;
+                break;
+            }
+            emit ValidatorRemoved(_validator);
+        }
+        require(validators.length>=reqValidatorCnt,"Validator size less than required count after removal");
```

Comment:
we can shift this before removal with
        require(validators.length-1>=reqValidatorCnt,"Validator size less than required count after removal");

Reply:
updated